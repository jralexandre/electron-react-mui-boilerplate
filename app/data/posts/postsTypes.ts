import { Action } from 'redux';

export enum PostActionTypes {
    ADD_POST = 'ADD_POST',
    EDIT_POST = 'EDIT_POST',
    DELETE_POST = 'DELETE_POST'
}

export interface AddPostAction extends Action<PostActionTypes.ADD_POST> {
    data: PostType;
}

export interface EditPostAction extends Action<PostActionTypes.EDIT_POST> {
    id: number;
    data: PostType;
}

export interface DeletePostAction extends Action<PostActionTypes.DELETE_POST> {
    id: number;
}

export type CustomPostAction = Action<''>;

export type PostAction =
    | AddPostAction
    | EditPostAction
    | DeletePostAction
    | CustomPostAction;

export interface PostType {
    id: number;
    title: string;
    content: string;
}

export interface PostsType {
    byId: {
        [id: number]: PostType;
    };
    allIds: number[];
    newPostId: number;
}
