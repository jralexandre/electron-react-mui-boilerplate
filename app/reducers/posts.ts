import {
    PostAction,
    ADD_POST,
    AddPostAction,
    EDIT_POST,
    EditPostAction,
    DELETE_POST
} from '../actions/posts';

export interface Post {
    id: number;
    title: string;
    content: string;
}

export interface Posts {
    byId: {
        [id: number]: Post;
    };
    allIds: number[];
    newPostId: number;
}

export default function posts(
    state: Posts = {
        byId: {},
        allIds: [],
        newPostId: 1
    },
    action: PostAction
): Posts {
    switch (action.type) {
        case ADD_POST:
            const newId = state.newPostId;
            return {
                byId: {
                    ...state.byId,
                    [newId]: {
                        ...(action as AddPostAction).data,
                        id: newId
                    }
                },
                allIds: state.allIds.concat(newId),
                newPostId: newId + 1
            };
        case EDIT_POST:
            const ac = action as EditPostAction;
            if (!state.allIds.includes(ac.id)) return state;

            return {
                byId: {
                    ...state.byId,
                    [ac.id]: { ...ac.data }
                },
                allIds: state.allIds,
                newPostId: state.newPostId
            };
        case DELETE_POST:
            if (!state.allIds.includes(action.id)) return state;

            const newById = state.byId;
            delete newById[action.id];

            return {
                byId: newById,
                allIds: state.allIds.filter(i => i !== action.id),
                newPostId: state.newPostId
            };
        default:
            return state;
    }
}
