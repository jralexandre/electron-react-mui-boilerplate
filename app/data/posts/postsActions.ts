import { ActionCreator, ActionCreatorsMapObject } from 'redux';
import { PostAction, Post, PostActionTypes } from './postsTypes';

export const addPost: ActionCreator<PostAction> = (data: Post) => ({
    type: PostActionTypes.ADD_POST,
    data
});

export const editPost: ActionCreator<PostAction> = (
    id: number,
    data: Post
) => ({
    type: PostActionTypes.EDIT_POST,
    id,
    data
});

export const deletePost: ActionCreator<PostAction> = (id: number) => ({
    type: PostActionTypes.DELETE_POST,
    id
});

export const postsActionCreators: ActionCreatorsMapObject<PostAction> = {
    addPost,
    editPost,
    deletePost
};
