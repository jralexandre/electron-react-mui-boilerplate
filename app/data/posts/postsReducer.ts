import { PostsType, PostAction, PostActionTypes } from './postsTypes';

export default function posts(
    state: PostsType = {
        byId: {
            1: { id: 1, title: 'Sample post', content: 'This is a sample post.' }
        },
        allIds: [ 1 ],
        newPostId: 2
    },
    action: PostAction
): PostsType {
    switch (action.type) {
        case PostActionTypes.ADD_POST:
            return {
                byId: {
                    ...state.byId,
                    [state.newPostId]: {
                        ...action.data,
                        id: state.newPostId
                    }
                },
                allIds: state.allIds.concat(state.newPostId),
                newPostId: state.newPostId + 1
            };
        case PostActionTypes.EDIT_POST:
            return state.allIds.includes(action.id)
                ? {
                      byId: {
                          ...state.byId,
                          [action.id]: { ...action.data, id: action.id }
                      },
                      allIds: state.allIds,
                      newPostId: state.newPostId
                  }
                : state;
        case PostActionTypes.DELETE_POST:
            if (!state.allIds.includes(action.id)) return state;

            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const { [action.id]: _, ...newById } = state.byId;

            return {
                byId: newById,
                allIds: state.allIds.filter(i => i !== action.id),
                newPostId: state.newPostId
            };
        default:
            return state;
    }
}
