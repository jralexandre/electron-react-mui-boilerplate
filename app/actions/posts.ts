import { ActionCreator, Action, ActionCreatorsMapObject } from 'redux';
import { Post } from '../reducers/posts';

export const ADD_POST = 'ADD_POST';
export const EDIT_POST = 'EDIT_POST';
export const DELETE_POST = 'DELETE_POST';

export interface AddPostAction extends Action<typeof ADD_POST> {
    data: Post;
}

export interface EditPostAction extends Action<typeof EDIT_POST> {
    id: number;
    data: Post;
}

export interface DeletePostAction extends Action<typeof DELETE_POST> {
    id: number;
}

export type PostAction = AddPostAction | EditPostAction | DeletePostAction;

export const addPost: ActionCreator<AddPostAction> = (data: Post) => ({
    type: ADD_POST,
    data
});

export const editPost: ActionCreator<EditPostAction> = (
    id: number,
    data: Post
) => ({
    type: EDIT_POST,
    id,
    data
});

export const deletePost: ActionCreator<DeletePostAction> = (id: number) => ({
    type: DELETE_POST,
    id
});

export const postsActionCreators: ActionCreatorsMapObject<PostAction> = {
    addPost,
    editPost,
    deletePost
};
