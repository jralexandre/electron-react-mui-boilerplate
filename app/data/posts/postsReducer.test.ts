import posts from './postsReducer';
import { PostsType, PostActionTypes } from './postsTypes';

describe('postsReducer', () => {
    const sampleState: PostsType = {
        byId: {
            1: { id: 1, title: 'Sample Post', content: "Sample post's content" }
        },
        allIds: [1],
        newPostId: 2
    };

    it('should handle ADD_POST', () => {
        const newState = posts(sampleState, {
            type: PostActionTypes.ADD_POST,
            data: { id: -1, title: 'Added post', content: 'Dummy' }
        });

        expect(newState.allIds).toHaveLength(2);
        expect(newState.allIds).toContain(2);
        expect(newState.byId[2]).toEqual({
            id: 2,
            title: 'Added post',
            content: 'Dummy'
        });
        expect(newState.newPostId).toBe(3);
    });

    it('should handle EDIT_POST', () => {
        const newState = posts(sampleState, {
            type: PostActionTypes.EDIT_POST,
            id: 1,
            data: { id: 2, title: 'Edited post', content: 'Edited content' }
        });

        expect(newState.allIds).toHaveLength(1);
        expect(newState.byId[1]).toEqual({
            id: 1, // id should not be changed
            title: 'Edited post',
            content: 'Edited content'
        });
    });

    it('should handle DELETE_POST', () => {
        const newState = posts(sampleState, {
            type: PostActionTypes.DELETE_POST,
            id: 1
        });

        expect(newState.allIds).toHaveLength(0);
        expect(Object.entries(newState.byId)).toHaveLength(0);
    });

    it('should handle custom action', () => {
        const newState = posts(sampleState, { type: '' });

        expect(newState).toEqual(sampleState);
    });

    it('should handle initial state', () => {
        expect(posts(undefined, { type: '' })).toMatchSnapshot();
    });

    it('should not edit non existing post', () => {
        const newState = posts(sampleState, {
            type: PostActionTypes.EDIT_POST,
            id: 2,
            data: { id: 2, title: 'Invalid', content: 'Invalid' }
        });

        expect(newState.allIds).not.toContain(2);
        expect(Object.values(newState.byId)).not.toContain(
            expect.objectContaining({ title: 'Invalid' })
        );
    });

    it('should not delete non existing post', () => {
        const newState = posts(sampleState, {
            type: PostActionTypes.DELETE_POST,
            id: 2
        });

        expect(newState.allIds).toHaveLength(1);
        expect(newState.byId[1]).not.toBeUndefined();
    });
});
