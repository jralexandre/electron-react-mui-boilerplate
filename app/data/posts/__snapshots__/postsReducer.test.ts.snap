// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postsReducer should handle initial state 1`] = `
Object {
  "allIds": Array [
    1,
  ],
  "byId": Object {
    "1": Object {
      "content": "This is a sample post.",
      "id": 1,
      "title": "Sample post",
    },
  },
  "newPostId": 2,
}
`;
